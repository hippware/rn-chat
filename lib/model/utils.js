"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const mobx_1 = require("mobx");
function waitFor(condition) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => {
            mobx_1.when(() => {
                let res = false;
                try {
                    res = condition();
                }
                catch (e) {
                    reject(e);
                }
                return res;
            }, () => {
                resolve();
            });
        });
    });
}
exports.waitFor = waitFor;
//# sourceMappingURL=utils.js.map