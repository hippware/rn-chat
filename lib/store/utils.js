"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const mobx_1 = require("mobx");
function pad(n, width, z = '0') {
    const str = n.toString();
    return str.length >= width ? str : new Array(width - str.length + 1).join(z) + n;
}
function process(result) {
    if (typeof result === 'object') {
        if (Array.isArray(result)) {
            return result.map(el => process(el));
        }
        else {
            if (result['#text'] && Object.keys(result).length === 1) {
                return result['#text'];
            }
            const res = {};
            let changed = false;
            // eslint-disable-next-line
            for (const key in result) {
                // eslint-disable-next-line
                if (result.hasOwnProperty(key)) {
                    changed = true;
                    res[key] = process(result[key]);
                }
            }
            if (!changed) {
                return undefined;
            }
            return res;
        }
    }
    else {
        return result;
    }
}
function isArray(res) {
    return Array.isArray(res) || mobx_1.isObservableArray(res);
}
exports.isArray = isArray;
function camelize(str) {
    return str
        .replace(/\W|_|\d/g, ' ')
        .replace(/(?:^\w|[A-Z]|\b\w)/g, (letter, index) => {
        return index === 0 ? letter.toLowerCase() : letter.toUpperCase();
    })
        .replace(/\s+/g, '');
}
exports.camelize = camelize;
function processMap(data) {
    const res = {};
    Object.keys(data).forEach(key => {
        const value = data[key];
        try {
            if (value && value !== 'null' && key !== 'field') {
                if (key === 'roles') {
                    res.roles = isArray(data.roles.role) ? data.roles.role : [data.roles.role];
                }
                else if (['followers', 'bots', 'followed'].indexOf(key) !== -1) {
                    res[key + 'Size'] = parseInt(data[key].size);
                }
                else if (data[key].thumbnail_url !== undefined) {
                    // we have image here!
                    if (data[key]['#text']) {
                        res[key] = { id: data[key]['#text'], url: data[key].thumbnail_url };
                    }
                }
                else if (key === 'subscribed') {
                    res.isSubscribed = value === 'true';
                }
                else if (key === 'owner') {
                    res.owner = Strophe.getNodeFromJid(value);
                }
                else if (key === 'subscribers') {
                    res.followersSize = parseInt(value.size);
                }
                else if (key === 'location') {
                    res.location = { latitude: parseFloat(value.geoloc.lat), longitude: parseFloat(value.geoloc.lon) };
                }
                else if (key === 'updated') {
                    res.time = iso8601toDate(value).getTime();
                }
                else if (key === 'radius') {
                    res.radius = parseFloat(value);
                }
                else {
                    const numbers = ['image_items', 'total_items', 'visibility'];
                    res[camelize(key)] = numbers.indexOf(key) !== -1 ? parseInt(value) : value;
                }
            }
        }
        catch (e) {
            console.error(`Cannot process key ${key} value: ${value}`);
        }
    });
    delete res.id;
    return res;
}
exports.processMap = processMap;
function fromCamelCase(data = {}) {
    const { firstName, userID, phoneNumber, lastName, sessionID, uuid } = data, result = tslib_1.__rest(data, ["firstName", "userID", "phoneNumber", "lastName", "sessionID", "uuid"]);
    if (phoneNumber) {
        result.phone_number = phoneNumber;
        result.phoneNumber = phoneNumber;
    }
    if (userID) {
        result.auth_user = userID;
    }
    if (firstName) {
        result.first_name = firstName;
    }
    if (lastName) {
        result.last_name = lastName;
    }
    if (sessionID) {
        result.token = sessionID;
    }
    if (uuid) {
        result.user = uuid;
    }
    return result;
}
exports.fromCamelCase = fromCamelCase;
function getJid(username, host, resource) {
    let jid = `${username}@${host}`;
    if (resource) {
        jid = `${jid}/${resource}`;
    }
    return jid;
}
exports.getJid = getJid;
function getNodeJid(jid) {
    if (jid.indexOf('@') < 0) {
        return jid;
    }
    return jid.split('@')[0];
}
exports.getNodeJid = getNodeJid;
/** Function: getUniqueId
 *  Generate a unique ID for use in <iq/> elements.
 *
 *  All <iq/> stanzas are required to have unique id attributes.  This
 *  function makes creating these easy.  Each connection instance has
 *  a counter which starts from zero, and the value of this counter
 *  plus a colon followed by the suffix becomes the unique id. If no
 *  suffix is supplied, the counter is used as the unique id.
 *
 *  Suffixes are used to make debugging easier when reading the stream
 *  data, and their use is recommended.  The counter resets to 0 for
 *  every new connection for the same reason.  For connections to the
 *  same server that authenticate the same way, all the ids should be
 *  the same, which makes it easy to see changes.  This is useful for
 *  automated testing as well.
 *
 *  Parameters:
 *    (String) suffix - A optional suffix to append to the id.
 *
 *  Returns:
 *    A unique string to be used for the id attribute.
 */
function getUniqueId(suffix) {
    let uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
        let r = (Math.random() * 16) | 0, // eslint-disable-line
        v = c === 'x' ? r : (r & 0x3) | 0x8; // eslint-disable-line
        return v.toString(16);
    });
    if (typeof suffix === 'string' || typeof suffix === 'number') {
        return `${uuid}:${suffix}`;
    }
    else {
        return `${uuid}`;
    }
}
exports.getUniqueId = getUniqueId;
function generateID() {
    const time = Date.now();
    return `s${time}${pad(Math.round(Math.random() * 1000), 4)}`;
}
exports.generateID = generateID;
function parseXml(xml, arrayTags) {
    function isArray(o) {
        return Object.prototype.toString.apply(o) === '[object Array]';
    }
    function parseNode(xmlNode, res) {
        if (!xmlNode) {
            return;
        }
        if (xmlNode.nodeName === '#text') {
            /* if you want the object to have a properyty "#text" even if it is "",
                               remove that if-else and use code that is currently in else block
                               */
            if (xmlNode.nodeValue.trim() === '') {
                return;
            }
            else {
                res[xmlNode.nodeName] = xmlNode.nodeValue;
                return;
            }
        }
        const jsonNode = {};
        const existing = res[xmlNode.nodeName];
        if (existing) {
            if (!isArray(existing)) {
                res[xmlNode.nodeName] = [existing, jsonNode];
            }
            else {
                res[xmlNode.nodeName].push(jsonNode);
            }
        }
        else if (arrayTags && arrayTags.indexOf(xmlNode.nodeName) !== -1) {
            res[xmlNode.nodeName] = [jsonNode];
        }
        else {
            res[xmlNode.nodeName] = jsonNode;
        }
        if (xmlNode.attributes) {
            const len = xmlNode.attributes.length;
            for (let i = 0; i < len; i++) {
                const attribute = xmlNode.attributes[i];
                jsonNode[attribute.nodeName] = attribute.nodeValue;
            }
        }
        const length = xmlNode.childNodes.length;
        for (let i = 0; i < length; i++) {
            parseNode(xmlNode.childNodes[i], jsonNode);
        }
    }
    const result = {};
    parseNode(xml, result);
    return process(result);
}
exports.parseXml = parseXml;
// return hashcode for given string
function hashCode(s) {
    let hash = 0, i, chr, len;
    if (s.length === 0)
        return hash;
    for (i = 0, len = s.length; i < len; i++) {
        chr = s.charCodeAt(i);
        hash = (hash << 5) - hash + chr; // eslint-disable-line
        // Convert to 32bit integer.
        hash = Math.abs(hash & hash); // eslint-disable-line
    }
    return hash;
}
exports.hashCode = hashCode;
function iso8601toDate(date) {
    let timestamp = Date.parse(date), minutesOffset = 0;
    if (isNaN(timestamp)) {
        // eslint-disable-next-line
        let struct = /^(\d{4}|[+\-]\d{6})-(\d{2})-(\d{2})(?:[T ](\d{2}):(\d{2})(?::(\d{2})(?:\.(\d{3,}))?)?(?:(Z)|([+\-])(\d{2})(?::?(\d{2}))?))?/.exec(date);
        if (struct) {
            if (struct[8] !== 'Z') {
                minutesOffset = +struct[10] * 60 + +struct[11];
                if (struct[9] === '+') {
                    minutesOffset = -minutesOffset;
                }
            }
            return new Date(+struct[1], +struct[2] - 1, +struct[3], +struct[4], +struct[5] + minutesOffset, +struct[6], struct[7] ? +struct[7].substr(0, 3) : 0);
        }
        else {
            // XEP-0091 dateAsString
            timestamp = Date.parse(`${date.replace(/^(\d{4})(\d{2})(\d{2})/, '$1-$2-$3')}Z`);
        }
    }
    return new Date(timestamp);
}
exports.iso8601toDate = iso8601toDate;
//# sourceMappingURL=utils.js.map