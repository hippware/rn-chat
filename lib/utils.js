"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
function process(result) {
    if (typeof result === 'object') {
        if (Array.isArray(result)) {
            return result.map(function (el) { return process(el); });
        }
        else {
            if (result['#text'] && Object.keys(result).length === 1) {
                return result['#text'];
            }
            var res = {};
            var changed = false;
            // eslint-disable-next-line
            for (var key in result) {
                // eslint-disable-next-line
                if (result.hasOwnProperty(key)) {
                    changed = true;
                    res[key] = process(result[key]);
                }
            }
            if (!changed) {
                return undefined;
            }
            return res;
        }
    }
    else {
        return result;
    }
}
exports.default = {
    fromCamelCase: function (data) {
        if (data === void 0) { data = {}; }
        var firstName = data.firstName, userID = data.userID, phoneNumber = data.phoneNumber, lastName = data.lastName, sessionID = data.sessionID, uuid = data.uuid, result = tslib_1.__rest(data, ["firstName", "userID", "phoneNumber", "lastName", "sessionID", "uuid"]);
        if (phoneNumber) {
            result.phone_number = phoneNumber;
            result.phoneNumber = phoneNumber;
        }
        if (userID) {
            result.auth_user = userID;
        }
        if (firstName) {
            result.first_name = firstName;
        }
        if (lastName) {
            result.last_name = lastName;
        }
        if (sessionID) {
            result.token = sessionID;
        }
        if (uuid) {
            result.user = uuid;
        }
        return result;
    },
    getJid: function (username, host, resource) {
        var jid = username + "@" + host;
        if (resource) {
            jid = jid + "/" + resource;
        }
        return jid;
    },
    getNodeJid: function (jid) {
        if (jid.indexOf('@') < 0) {
            return null;
        }
        return jid.split('@')[0];
    },
    /** Function: getUniqueId
     *  Generate a unique ID for use in <iq/> elements.
     *
     *  All <iq/> stanzas are required to have unique id attributes.  This
     *  function makes creating these easy.  Each connection instance has
     *  a counter which starts from zero, and the value of this counter
     *  plus a colon followed by the suffix becomes the unique id. If no
     *  suffix is supplied, the counter is used as the unique id.
     *
     *  Suffixes are used to make debugging easier when reading the stream
     *  data, and their use is recommended.  The counter resets to 0 for
     *  every new connection for the same reason.  For connections to the
     *  same server that authenticate the same way, all the ids should be
     *  the same, which makes it easy to see changes.  This is useful for
     *  automated testing as well.
     *
     *  Parameters:
     *    (String) suffix - A optional suffix to append to the id.
     *
     *  Returns:
     *    A unique string to be used for the id attribute.
     */
    getUniqueId: function (suffix) {
        var uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            var r = (Math.random() * 16) | 0, // eslint-disable-line
            v = c === 'x' ? r : (r & 0x3) | 0x8; // eslint-disable-line
            return v.toString(16);
        });
        if (typeof suffix === 'string' || typeof suffix === 'number') {
            return uuid + ":" + suffix;
        }
        else {
            return "" + uuid;
        }
    },
    parseXml: function (xml, arrayTags) {
        function isArray(o) {
            return Object.prototype.toString.apply(o) === '[object Array]';
        }
        function parseNode(xmlNode, res) {
            if (!xmlNode) {
                return;
            }
            if (xmlNode.nodeName === '#text') {
                /* if you want the object to have a properyty "#text" even if it is "",
                                 remove that if-else and use code that is currently in else block
                                 */
                if (xmlNode.nodeValue.trim() === '') {
                    return;
                }
                else {
                    res[xmlNode.nodeName] = xmlNode.nodeValue;
                    return;
                }
            }
            var jsonNode = {};
            var existing = res[xmlNode.nodeName];
            if (existing) {
                if (!isArray(existing)) {
                    res[xmlNode.nodeName] = [existing, jsonNode];
                }
                else {
                    res[xmlNode.nodeName].push(jsonNode);
                }
            }
            else if (arrayTags && arrayTags.indexOf(xmlNode.nodeName) !== -1) {
                res[xmlNode.nodeName] = [jsonNode];
            }
            else {
                res[xmlNode.nodeName] = jsonNode;
            }
            if (xmlNode.attributes) {
                var len = xmlNode.attributes.length;
                for (var i = 0; i < len; i++) {
                    var attribute = xmlNode.attributes[i];
                    jsonNode[attribute.nodeName] = attribute.nodeValue;
                }
            }
            var length = xmlNode.childNodes.length;
            for (var i = 0; i < length; i++) {
                parseNode(xmlNode.childNodes[i], jsonNode);
            }
        }
        var result = {};
        parseNode(xml, result);
        return process(result);
    },
    // return hashcode for given string
    hashCode: function (s) {
        var hash = 0, i, chr, len;
        if (s.length === 0)
            return hash;
        for (i = 0, len = s.length; i < len; i++) {
            chr = s.charCodeAt(i);
            hash = (hash << 5) - hash + chr; // eslint-disable-line
            // Convert to 32bit integer.
            hash = Math.abs(hash & hash); // eslint-disable-line
        }
        return hash;
    },
    iso8601toDate: function (date) {
        var timestamp = Date.parse(date), minutesOffset = 0;
        if (isNaN(timestamp)) {
            // eslint-disable-next-line
            var struct = /^(\d{4}|[+\-]\d{6})-(\d{2})-(\d{2})(?:[T ](\d{2}):(\d{2})(?::(\d{2})(?:\.(\d{3,}))?)?(?:(Z)|([+\-])(\d{2})(?::?(\d{2}))?))?/.exec(date);
            if (struct) {
                if (struct[8] !== 'Z') {
                    minutesOffset = +struct[10] * 60 + +struct[11];
                    if (struct[9] === '+') {
                        minutesOffset = -minutesOffset;
                    }
                }
                return new Date(+struct[1], +struct[2] - 1, +struct[3], +struct[4], +struct[5] + minutesOffset, +struct[6], struct[7] ? +struct[7].substr(0, 3) : 0);
            }
            else {
                // XEP-0091 dateAsString
                timestamp = Date.parse(date.replace(/^(\d{4})(\d{2})(\d{2})/, '$1-$2-$3') + "Z");
            }
        }
        return new Date(timestamp);
    }
};
//# sourceMappingURL=utils.js.map