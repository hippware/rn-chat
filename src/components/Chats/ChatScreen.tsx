import React from 'react'
import {
  View,
  Keyboard,
  Text,
  TouchableOpacity,
  Image,
  StyleSheet,
  ActivityIndicator,
  FlatList,
} from 'react-native'

import moment from 'moment'
import {observable} from 'mobx'
import {observer, inject} from 'mobx-react/native'
import {isAlive} from 'mobx-state-tree'
import {Actions} from 'react-native-router-flux'
import Screen from '../Screen'
import {showImagePicker} from '../ImagePicker'
import ChatBubble from './ChatBubble'
import ChatMessage from './ChatMessage'
import {AutoExpandingTextInput, Avatar} from '../common'
import {colors} from '../../constants'
import {IWocky} from 'wocky-client'

const Button = require('apsl-react-native-button')

type Props = {
  item: string
  wocky?: IWocky
}

type State = {
  text: string
  height: number
}

const ChatTitle = inject('wocky')(
  observer(({item, wocky}) => {
    return wocky.chats.get(item)
      ? wocky.chats.get(item).participants.map((profile, ind) => (
          <TouchableOpacity
            key={`${ind}${profile.id}touch`} // eslint-disable-line
            onPress={() => {
              Actions.profileDetail({item: profile, title: profile.displayName})
            }}
          >
            <Avatar size={40} profile={profile} />
          </TouchableOpacity>
        ))
      : null
  })
)

@inject('wocky')
@observer
class ChatScreen extends React.Component<Props, State> {
  static renderTitle = ({item}) => <ChatTitle item={item} />

  state: State = {
    text: '',
    height: 0,
  }

  @observable messages: any[] = []
  @observable chat: any
  mounted: boolean = false
  handler: any
  list: any

  componentDidMount() {
    const {item, wocky} = this.props
    this.chat = wocky!.createChat(item)
    // load chat asynchronously
    this.chat.load().then(() => this.chat.readAll())
    this.chat.setActive(true)
    Keyboard.addListener('keyboardWillShow', this.keyboardWillShow)
    Keyboard.addListener('keyboardWillHide', this.keyboardWillHide)
    this.mounted = true
  }

  componentWillUnmount() {
    this.mounted = false
    if (this.chat) {
      this.chat.setActive(false)
    }
    Keyboard.removeListener('keyboardWillShow', this.keyboardWillShow)
    Keyboard.removeListener('keyboardWillHide', this.keyboardWillHide)
    if (this.handler) {
      this.handler()
      this.handler = null
    }
  }

  onSend = () => {
    if (this.chat.message.body.trim()) {
      this.chat.message.send()
    }
  }

  keyboardWillShow = e => {
    if (this.mounted) this.setState({height: e.endCoordinates.height})
  }

  keyboardWillHide = () => {
    if (this.mounted) this.setState({height: 0})
  }

  // TODO: rework this so it's included in row render...inefficient this way
  renderDate = (rowData: any = {}) => {
    let diffMessage: any = null
    diffMessage = this.getPreviousMessage(rowData)
    if (rowData.date instanceof Date) {
      if (diffMessage === null) {
        return <Text style={[styles.date]}>{moment(rowData.date).calendar()}</Text>
      } else if (diffMessage.date instanceof Date) {
        const diff = moment(rowData.date).diff(diffMessage.date, 'minutes')
        if (diff > 5) {
          return <Text style={[styles.date]}>{moment(rowData.date).calendar()}</Text>
        }
      }
    }
    return null
  }

  getPreviousMessage = message => {
    const i = this.messages.findIndex(m => m.uniqueId === message.uniqueId)
    return this.messages.length > i + 1 ? this.messages[i + 1] : null
  }

  renderItem = ({item}) =>
    item ? (
      <View>
        {this.renderDate(item)}
        <ChatMessage
          rowData={item}
          diffMessage={this.getPreviousMessage(item)}
          position={item.position}
        />
      </View>
    ) : null

  _footerComponent: any = observer(
    () =>
      this.chat && this.chat.loading ? <ActivityIndicator style={{marginVertical: 20}} /> : null
  )

  render() {
    return this.chat && isAlive(this.chat) ? (
      <Screen>
        <FlatList
          inverted
          data={this.chat.messages
            .map(el => {
              let media = null
              try {
                media = el.media
              } catch (err) {
                // console.log('TODO: fix Message.media reference error', err)
              }
              return el
                ? {
                    uniqueId: el.id,
                    text: el.body || '',
                    isDay: true,
                    title: el.from.displayName,
                    media,
                    size: 40,
                    position: el.from.isOwn ? 'right' : 'left',
                    status: '',
                    name: el.from.isOwn ? '' : el.from.displayName,
                    image:
                      el.from.isOwn || !el.from.avatar || !el.from.avatar.source
                        ? null
                        : el.from.avatar.source,
                    profile: el.from,
                    imageView: Avatar,
                    view: ChatBubble,
                    date: new Date(el.time),
                  }
                : null
            })
            .reverse()}
          ref={l => (this.list = l)}
          renderItem={this.renderItem}
          keyExtractor={i => i.uniqueId}
          onEndReached={() => this.chat.load()}
          onEndReachedThreshold={0.5}
          ListFooterComponent={this._footerComponent}
        />
        <InputArea chat={this.chat} onSend={this.onSend} />
        <View style={{height: this.state.height}} />
      </Screen>
    ) : (
      <Screen />
    )
  }
}

// separating text input here prevents unnecessary re-renders of the entire list when user enters text
const InputArea = inject('wocky')(
  observer(({wocky, chat, onSend}) => {
    return chat.message ? (
      <View style={[styles.textInputContainer, styles.textInputContainerDay]}>
        <AttachButton message={chat.message} />
        <AutoExpandingTextInput
          style={[styles.textInput, styles.textInputDay]}
          placeholder="Write a message"
          placeholderTextColor={colors.DARK_GREY}
          multiline
          autoFocus
          returnKeyType="default"
          enablesReturnKeyAutomatically
          onChangeText={t => chat.message && chat.message.setBody(t)}
          value={chat.message.body}
          blurOnSubmit={false}
          maxHeight={100}
          maxLength={500}
        />
        <TouchableOpacity disabled={!chat.message.body.trim() || !wocky.connected} onPress={onSend}>
          <Image
            source={
              chat.message.body.trim() && wocky.connected
                ? require('../../../images/iconSendActive.png')
                : require('../../../images/iconSendInactive.png')
            }
          />
        </TouchableOpacity>
      </View>
    ) : null
  })
)

const onAttach = (message, notificationStore) => {
  showImagePicker({
    title: 'Select Image',
    callback: async (source, response) => {
      try {
        await message.upload({
          file: source,
          width: response.width,
          height: response.height,
          size: response.size,
        })
        message.send()
      } catch (e) {
        notificationStore.flash(e.message)
      }
    },
  })
}

const AttachButton = inject('notificationStore')(({notificationStore, message}) => (
  <Button
    style={{borderWidth: 0, borderColor: 'transparent', paddingTop: 4}}
    onPress={() => onAttach(message, notificationStore)}
  >
    <Image source={require('../../../images/iconAttach.png')} />
  </Button>
))

export default ChatScreen

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'transparent',
  },
  listView: {
    flex: 1,
  },
  textInputContainerDay: {
    backgroundColor: 'white',
  },
  textInputContainerNight: {
    backgroundColor: 'rgba(63,50,77,0.9)',
  },
  textInputContainer: {
    flexDirection: 'row',
    paddingHorizontal: 20,
    alignItems: 'center',
    justifyContent: 'center',
  },
  textInput: {
    width: 100,
    fontFamily: 'Roboto-Regular',
    flex: 1,
    margin: 0,
    padding: 0,
    paddingLeft: 20,
    paddingRight: 20,
    fontSize: 15,
  },
  textInputDay: {
    color: colors.DARK_PURPLE,
  },
  textInputNight: {
    color: 'white',
  },
  date: {
    color: '#aaaaaa',
    fontSize: 12,
    textAlign: 'center',
    fontWeight: 'bold',
    marginBottom: 8,
  },
  link: {
    color: '#007aff',
    textDecorationLine: 'underline',
  },
  linkLeft: {
    color: '#000',
  },
  linkRight: {
    color: '#fff',
  },
  loadEarlierMessages: {
    height: 44,
    justifyContent: 'center',
    alignItems: 'center',
  },
  loadEarlierMessagesButton: {
    fontSize: 14,
  },
})
