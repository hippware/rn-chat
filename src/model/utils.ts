import {when} from 'mobx'

export async function waitFor(condition: () => boolean) {
  return new Promise((resolve, reject) => {
    when(
      () => {
        let res = false
        try {
          res = condition()
        } catch (e) {
          reject(e)
        }
        return res
      },
      () => {
        resolve()
      }
    )
  })
}
