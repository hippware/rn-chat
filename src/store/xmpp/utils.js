// pad given number with given width and symbol z, like (1, 4, '0') -> '0001'
function pad (n, width, z) {
    z = z || '0'
    n = n + ''
    return n.length >= width ? n : new Array(width - n.length + 1).join(z) + n
}

function process (result) {
    if (typeof result === 'object') {
        if (Array.isArray(result)) {
            return result.map(el => process(el))
        } else {
            if (result['#text'] && Object.keys(result).length === 1) {
                return result['#text']
            }
            let res = {}
            let changed = false
            for (let key in result) {
                if (result.hasOwnProperty(key)) {
                    changed = true
                    res[key] = process(result[key])
                }
            }
            if (!changed) {
                return undefined
            }
            return res
        }
    } else {
        return result
    }

}

export default {
    clone(obj) {
        if (Array.isArray(obj)) {
            return obj.map(x => clone(x))
        }
        if (obj == null || typeof(obj) != 'object') {
            return obj
        }
        var temp = new obj.constructor()

        for (var key in obj) {

            if (obj.hasOwnProperty(key)) {
                if (key != 'state') {
                    temp[key] = clone(obj[key])
                }
            }
        }

        return temp
    },
    getJid(username, host, resource){
        let jid = username + '@' + host
        if (resource) {
            jid = jid + '/' + resource
        }
        return jid
    },
    getNodeJid(jid) {
        if (jid.indexOf('@') < 0) {
            return null
        }
        return jid.split('@')[0]
    },

    /** Function: getUniqueId
     *  Generate a unique ID for use in <iq/> elements.
     *
     *  All <iq/> stanzas are required to have unique id attributes.  This
     *  function makes creating these easy.  Each connection instance has
     *  a counter which starts from zero, and the value of this counter
     *  plus a colon followed by the suffix becomes the unique id. If no
     *  suffix is supplied, the counter is used as the unique id.
     *
     *  Suffixes are used to make debugging easier when reading the stream
     *  data, and their use is recommended.  The counter resets to 0 for
     *  every new connection for the same reason.  For connections to the
     *  same server that authenticate the same way, all the ids should be
     *  the same, which makes it easy to see changes.  This is useful for
     *  automated testing as well.
     *
     *  Parameters:
     *    (String) suffix - A optional suffix to append to the id.
     *
     *  Returns:
     *    A unique string to be used for the id attribute.
     */
    getUniqueId: function (suffix) {
        var uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            var r = Math.random() * 16 | 0,
                v = c == 'x' ? r : r & 0x3 | 0x8
            return v.toString(16)
        })
        if (typeof(suffix) == 'string' || typeof(suffix) == 'number') {
            return uuid + ':' + suffix
        } else {
            return uuid + ''
        }
    },

    parseXml(xml, arrayTags){
        function isArray (o) {
            return Object.prototype.toString.apply(o) === '[object Array]'
        }

        function parseNode (xmlNode, result) {
            if (!xmlNode) {
                return
            }
            if (xmlNode.nodeName == '#text') {
                /* if you want the object to have a properyty "#text" even if it is "",
                 remove that if-else and use code that is currently in else block
                 */
                if (xmlNode.nodeValue.trim() == '') {
                    return
                }
                else {
                    result[xmlNode.nodeName] = xmlNode.nodeValue
                    return
                }
            }

            var jsonNode = {}
            var existing = result[xmlNode.nodeName]
            if (existing) {
                if (!isArray(existing)) {
                    result[xmlNode.nodeName] = [existing, jsonNode]
                }
                else {
                    result[xmlNode.nodeName].push(jsonNode)
                }
            }
            else {
                if (arrayTags && arrayTags.indexOf(xmlNode.nodeName) != -1) {
                    result[xmlNode.nodeName] = [jsonNode]
                }
                else {
                    result[xmlNode.nodeName] = jsonNode
                }
            }

            if (xmlNode.attributes) {
                var length = xmlNode.attributes.length
                for (var i = 0; i < length; i++) {
                    var attribute = xmlNode.attributes[i]
                    jsonNode[attribute.nodeName] = attribute.nodeValue
                }
            }

            var length = xmlNode.childNodes.length
            for (var i = 0; i < length; i++) {
                parseNode(xmlNode.childNodes[i], jsonNode)
            }
        }

        var result = {}
        parseNode(xml, result)
        return process(result)
    },

    // generate ID for all xml stanzas
    generateID() {
        const time = Date.now()
        return `s${time}${pad(Math.round(Math.random() * 1000), 4)}`
    },

    // return hashcode for given string
    hashCode(s) {
        var hash = 0, i, chr, len
        if (s.length === 0) return hash
        for (i = 0, len = s.length; i < len; i++) {
            chr = s.charCodeAt(i)
            hash = ((hash << 5) - hash) + chr
            hash = Math.abs(hash & hash) // Convert to 32bit integer
        }
        console.log('HASH:', s, hash)

        return hash
    },

    iso8601toDate(date){
        var timestamp = Date.parse(date), minutesOffset = 0
        if (isNaN(timestamp)) {
            var struct = /^(\d{4}|[+\-]\d{6})-(\d{2})-(\d{2})(?:[T ](\d{2}):(\d{2})(?::(\d{2})(?:\.(\d{3,}))?)?(?:(Z)|([+\-])(\d{2})(?::?(\d{2}))?))?/.exec(date)
            if (struct) {
                if (struct[8] !== 'Z') {
                    minutesOffset = +struct[10] * 60 + (+struct[11])
                    if (struct[9] === '+') {
                        minutesOffset = -minutesOffset
                    }
                }
                return new Date(+struct[1], +struct[2] - 1, +struct[3], +struct[4], +struct[5] + minutesOffset, +struct[6], struct[7] ? +struct[7].substr(0, 3) : 0)
            } else {
                // XEP-0091 dateAsString
                timestamp = Date.parse(date.replace(/^(\d{4})(\d{2})(\d{2})/, '$1-$2-$3') + 'Z')
            }
        }
        return new Date(timestamp)
    }
}