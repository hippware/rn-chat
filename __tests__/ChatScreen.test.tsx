import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import ChatScreen from '../src/components/ChatScreen'
import './utils/mockTextInput'
import mockStore from './utils/mockStore'
import {Provider} from 'mobx-react/native'

describe('ChatScreen', () => {
  it('renders with no data', () => {
    const toRender = (
      <Provider {...mockStore}>
        <ChatScreen item="1234" />
      </Provider>
    )
    const tree = renderer.create(toRender).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
