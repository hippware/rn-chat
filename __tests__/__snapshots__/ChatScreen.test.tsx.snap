// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatScreen renders with no data 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
      "transform": Array [
        Object {
          "translateY": 0,
        },
      ],
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <BVLinearGradient
      colors={
        Array [
          "rgb(255,255,255)",
          "rgb(241,242,244)",
          "rgb(243,244,246)",
        ]
      }
      locations={
        Array [
          0,
          0.2,
          1,
        ]
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
        ]
      }
    >
      <RCTScrollView
        ListFooterComponent={[Function]}
        data={Array []}
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        invertStickyHeaders={true}
        inverted={true}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        numColumns={1}
        onContentSizeChange={[Function]}
        onEndReached={[Function]}
        onEndReachedThreshold={0.5}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View>
          <View
            onLayout={[Function]}
            style={
              Object {
                "transform": Array [
                  Object {
                    "scaleY": -1,
                  },
                ],
              }
            }
          />
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
